add styles to d3-cloud to fit container height and width and fix problem with ignored list

vaclavpavlicek opened the pull request:
I tried to use this library with angular material. I tried to render word cloud to md-grid-tile but size of d3-cloud was 0x0. So I updated styles. And I add waiting for document when it is ready, because without that svg had size 0x0. When you were checking ignored objects you tried to find indexOf object name in array of objects and it could work. So I created method for checking that. But I forget to remove comparing color. So in next pull request I'll add optional color of word.

Fixes #15, #16, #17, and #18

vaclavpavlicek committed:
add styles to d3-cloud to fit container height and width and fix problem with ignored words

grtjn commented:
Thanks for the PR, nice to see others being interested in this little project.

A few thoughts based on a first quick look:
- you can leave out the dist changes, I'll build on release..
- I am a little surprised it is necessary to wait for document ready, as angular itself is normally not initialized before document onready. This SO answer seems to confirm that: https://stackoverflow.com/a/29887821/918496
- I see you are trying to ignore words by properties other than just the name. I'm thinking that could be generalized. The directive could take an optional function to filter the words. A hard-coded default function that only looks at name could easily be applied.
- Or maybe even better: that (optional) filter function could replace ignoreList entirely. Use filter function if provided, otherwise use ignoreList with its current support.
- I like the idea of being able to pass in color along with each word, but it should be optional. You could potentially do things like: `(ignoredObject.color ? ignoredObject.color === newValue.color : true)`
- ` + 0` has dropped off twice around `var cloudHeight = $element[0].clientWidth`. I forgot its true purpose, but it probably had meaning.

Note: you can push more commits to your local 'commity' branch to update this PR..

vaclavpavlicek commented:
OK I'll add optional colors and then I'll send another PR.

grtjn commented:
No need for a new PR, just pile up commits in here..

vaclavpavlicek commented:
What do you think about filter function? Should it be something like filterFunction(value, ignore)? Both parameters are objects. Controller will send all combinations of values and ignores. Controller will get filter function from directive attribute. Is it OK?

grtjn commented:
I was thinking it only needs the value argument. You can provide ignore lists or whatever else is needed directly to the function before passing it into the directive..

vaclavpavlicek committed:
add optional color property

vaclavpavlicek committed:
replace ignored list of words with custom filter function

vaclavpavlicek commented:
I added optional color and custom filter function. Are they ok?

grtjn commented on dist/d3-cloud-ng.css:
```diff
@@ -1,3 +1,7 @@
+.d3-cloud {
+  height: 100%;
+  width: 100%;
+}
```
Please revert the dist changes

grtjn commented on src/d3-cloud.controller.js:
```diff
@@ -8,47 +8,43 @@
-                if (!updateflag & newValue[i].name !== oldValue[i].name & newValue[i].score !== oldValue[i].score) {
-                  updateflag = 1;
+    $scope.$watch('words', function (newValue, oldValue) {
+      angular.element(document).ready(function () {
```
Please remove this document.ready, or convince me for the need for this with a jsfiddle or such..

grtjn commented on src/d3-cloud.controller.js:
```diff
@@ -8,47 +8,43 @@
+        var updateflag = 0;
+
+        if (newValue) {
+          items = $scope.filterFunction($scope.words);
```
For consistency, please use `newValue` instead of `$scope.words`

grtjn replied:
Additionally, I think it would be nice if the filter function would be called for each word/item separately. It would take the burden away from people to iterate over the items themselves to filter them, and just focus on one item at a time. So basically restore the old code, but replace the `if (ignore.indexOf(newValue[i].name) < 1)` with `if (filter(newValue[i]))`..

grtjn commented on src/d3-cloud.controller.js:
```diff
@@ -8,47 +8,43 @@
+            if (oldValue) {
+              if (oldValue.length !== newValue.length) {
+                updateflag = 1;
+              } else if (items.length > 0) {
```
This `else if` should not be necessary. Actually, the `if` above and the `else` below should use `items` instead of `newValue`, making the `else if` obsolete..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
         events: '=?',
         font: '@',
-        ignoreList: '=?',
+        filterFunction: '=?',
```
Please preserve ignoreList for backwards compatibility. And I like short attribute names, so how about just `filter`? Could you also use the `&?` binding for the filter function?

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
-            .start();
-        };
+      link: function ($scope, $element, $attrs) {
+        angular.element(document).ready(function () {
```
Another document.ready which I am sure you should not need. Please remove..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
+          $scope.font = $scope.font || 'Impact';
+          $scope.filterFunction = $scope.filterFunction || function (words) {
+            return words;
+          };
```
The built-in filter function should not pass-through, but filter based on `ignoreList` (for backwards compatibility). I also think we should throw an exception if both are provided, and mark ignoreList as being DEPRECATED somehow..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
+
+          $scope.updateCloud = function (words) {
+
+            var cloudHeight = $element[0].clientHeight;
```
` + 0` dropped off, please restore..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
+          $scope.updateCloud = function (words) {
+
+            var cloudHeight = $element[0].clientHeight;
+            var cloudWidth = $element[0].clientWidth;
```
` + 0` dropped off, please restore..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
+            $scope.cloud = d3.layout.cloud().size([cloudWidth, cloudHeight]);
+            $scope.cloud
+              .words(words.map(function (d) {
+                const result = {
```
Don't use `const` here, just use `var`. Better legacy support, and result is not used as constant..

grtjn commented on src/d3-cloud.directive.js:
```diff
@@ -37,173 +37,191 @@
+            $scope.cloud = d3.layout.cloud().size([cloudWidth, cloudHeight]);
+            $scope.cloud
+              .words(words.map(function (d) {
+                const result = {
```
Don't use `const` here, just use `var`. Better legacy support, and result is not used as constant..

vaclavpavlicek committed:
add ignore list and marked it as deprecated

vaclavpavlicek commented:
I repaired filters are they ok? Setting width and height in styles and waiting for document ready is used due to angular material. Specifically due to using d3-cloud-ng in md-gridp-tile. It works fine in md-content, but in md-grid-tile it doesn't, because width and height of div with class d3-cloud is 19x0. When I added styles (height: 100% and width: 100%), div fit width and height of md-grid-tile but svg had width = 0 and height = 0 too. So when I added angular.element(document).ready directive waited for div to resize to 100% width and height 100% and then it created svg with size of div. Do you see what I mean?

vaclavpavlicek committed:
remove angular.element(document).ready and update styles

vaclavpavlicek commented:
I removed anguler.element(document).ready. Should I edit something else?

grtjn commented:
Awesome, looks pretty good. Only one tiny thing: the docs on the filter function are off. It still mentions a words array. How about:

    Optional. A function that filters words. Invoked for each word, should return true to retain the word, false to skip. Defaults to comparing against ignoreList for legacy purposes.

Regarding document.ready, I think I understand now, but I would think document.ready would not help here. It would sound more logical to me to wait for $element to be 'ready'. Now, I'm not sure you can use `$element.ready()`, but this answer seems to indicate a zero sec $timeout might do the trick: https://stackoverflow.com/a/32074491/918496. Maybe worth a look..

vaclavpavlicek committed:
update filter function documentation

vaclavpavlicek commented:
Documentation updated. Problem with rendering to md-grid-tile was solved with custom directive which broadcasts event on md-grid-list size count and then I show d3-cloud-ng.

vaclavpavlicek committed:
update filter function documentation

vaclavpavlicek commented:
Better?

grtjn commented:
I was about to suggest opening a separate ticket for the ready stuff, and you could still do so if you like.

Thanks for the doc update. I'll merge and tidy tomorrow if you don't mind, when I'm back behind the keyboard. I'll cut a new release so you can use your own work.. 😊

vaclavpavlicek commented:
I don't mind. :smile: Thank you very much! :smile:

grtjn commented:
I think it is useful to follow up on the document.ready stuff, but no need to have this PR wait for that. I created ticket #19 as a reminder to look into it at some point in future..

grtjn commented:
0.9.8 is released..

vaclavpavlicek commented:
Thank you! :smiley:
